File: proj.android_studio/ee-x-facebook-ads/src/main/java/com/ee/internal/FacebookRewardedAd.kt
package com.ee.internal

import android.content.Context
import androidx.annotation.AnyThread
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.Thread
import com.ee.Utils
import com.facebook.ads.Ad
import com.facebook.ads.AdError
import com.facebook.ads.RewardedVideoAd
import com.facebook.ads.RewardedVideoAdListener
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Created by Zinge on 10/11/17.
 */
internal class FacebookRewardedAd(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _context: Context,
    private val _adId: String) : RewardedVideoAdListener {
    companion object {
        private val kTag = FacebookRewardedAd::class.java.name
    }

    private val _messageHelper = MessageHelper("FacebookRewardedAd", _adId)
    private val _isLoaded = AtomicBoolean(false)
    private var _displaying = false
    private var _rewarded = false
    private var _ad: RewardedVideoAd? = null

    init {
        _logger.info("$kTag: constructor: adId = $_adId")
        registerHandlers()
        createInternalAd()
    }

    @AnyThread
    fun destroy() {
        _logger.info("$kTag: ${this::destroy.name}: adId = $_adId")
        deregisterHandlers()
        destroyInternalAd()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(_messageHelper.isLoaded) {
            Utils.toString(isLoaded)
        }
        _bridge.registerHandler(_messageHelper.load) {
            load()
            ""
        }
        _bridge.registerHandler(_messageHelper.show) {
            show()
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(_messageHelper.isLoaded)
        _bridge.deregisterHandler(_messageHelper.load)
        _bridge.deregisterHandler(_messageHelper.show)
    }

    @AnyThread
    private fun createInternalAd() {
        Thread.runOnMainThread {
            if (_ad != null) {
                return@runOnMainThread
            }
            _ad = RewardedVideoAd(_context, _adId)
        }
    }

    @AnyThread
    private fun destroyInternalAd() {
        Thread.runOnMainThread {
            val ad = _ad ?: return@runOnMainThread
            ad.destroy()
            _ad = null
        }
    }

    private val isLoaded: Boolean
        @AnyThread get() = _isLoaded.get()

    @AnyThread
    private fun load() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::load.name}")
            val ad = _ad ?: throw IllegalArgumentException("Ad is not initialized")
            ad.loadAd(ad.buildLoadAdConfig().withAdListener(this).build())
        }
    }

    @AnyThread
    private fun show() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::show.name}")
            val ad = _ad ?: throw IllegalArgumentException("Ad is not initialized")
            _displaying = true
            _rewarded = false
            val result = ad.show(ad.buildShowAdConfig().build())
            if (result) {
                // OK.
                _isLoaded.set(false)
            } else {
                destroyInternalAd()
                createInternalAd()
                _bridge.callCpp(_messageHelper.onFailedToShow)
            }
        }
    }

    override fun onError(ad: Ad, adError: AdError) {
        _logger.debug("$kTag: ${this::onError.name}: ${adError.errorMessage}")
        destroyInternalAd()
        createInternalAd()
        if (_displaying) {
            _displaying = false
            _bridge.callCpp(_messageHelper.onFailedToShow, adError.errorMessage)
        } else {
            _bridge.callCpp(_messageHelper.onFailedToLoad, adError.errorMessage)
        }
    }

    override fun onAdLoaded(ad: Ad) {
        _logger.debug("$kTag: ${this::onAdLoaded.name}")
        _isLoaded.set(true)
        _bridge.callCpp(_messageHelper.onLoaded)
    }

    override fun onLoggingImpression(ad: Ad) {
        _logger.debug("$kTag: ${this::onLoggingImpression.name}")
    }

    override fun onAdClicked(ad: Ad) {
        _logger.debug("$kTag: ${this::onAdClicked.name}")
        _bridge.callCpp(_messageHelper.onClicked)
    }

    override fun onRewardedVideoCompleted() {
        _logger.debug("$kTag: ${this::onRewardedVideoCompleted.name}")
        _rewarded = true
    }

    override fun onRewardedVideoClosed() {
        _logger.info(this::onRewardedVideoClosed.name)
        _displaying = false
        destroyInternalAd()
        createInternalAd()
        _bridge.callCpp(_messageHelper.onClosed, Utils.toString(_rewarded))
    }
}

File: src/ts/src/facebook_ads/internal/FacebookRewardedAd.ts
import {
    AdObserver,
    FullScreenAdResult,
    IFullScreenAd,
} from "../../ads";
import {
    AsyncHelper,
    IAsyncHelper,
    MessageHelper,
} from "../../ads/internal";
import {
    IMessageBridge,
    ObserverManager,
    Utils,
} from "../../core";
import { FacebookAds } from "./FacebookAds";

export class FacebookRewardedAd extends ObserverManager<AdObserver> implements IFullScreenAd {
    private readonly _bridge: IMessageBridge;
    private readonly _displayer: IAsyncHelper<FullScreenAdResult>;
    private readonly _plugin: FacebookAds;
    private readonly _adId: string;
    private readonly _messageHelper: MessageHelper;
    private _loadingCapped: boolean;
    private readonly _loader: IAsyncHelper<boolean>;

    public constructor(
        bridge: IMessageBridge, displayer: IAsyncHelper<FullScreenAdResult>, plugin: FacebookAds, adId: string) {
        super();
        this._bridge = bridge;
        this._displayer = displayer;
        this._plugin = plugin;
        this._adId = adId;
        this._messageHelper = new MessageHelper("FacebookRewardedAd", adId);
        this._loadingCapped = false;
        this._loader = new AsyncHelper<boolean>();

        this._bridge.registerHandler(_ => this.onLoaded(), this._messageHelper.onLoaded);
        this._bridge.registerHandler(this.onFailedToLoad, this._messageHelper.onFailedToLoad);
        this._bridge.registerHandler(this.onFailedToShow, this._messageHelper.onFailedToShow);
        this._bridge.registerHandler(_ => this.onClicked(), this._messageHelper.onClicked);
        this._bridge.registerHandler(message => this.onClosed(Utils.toBool(message)), this._messageHelper.onClosed);
    }

    public destroy(): void {
        this._bridge.deregisterHandler(this._messageHelper.onLoaded);
        this._bridge.deregisterHandler(this._messageHelper.onFailedToLoad);
        this._bridge.deregisterHandler(this._messageHelper.onFailedToShow);
        this._bridge.deregisterHandler(this._messageHelper.onClicked);
        this._bridge.deregisterHandler(this._messageHelper.onClosed);
        this._plugin.destroyRewardedAd(this._adId);
    }

    public get isLoaded(): boolean {
        const response = this._bridge.call(this._messageHelper.isLoaded);
        return Utils.toBool(response);
    }

    public async load(): Promise<boolean> {
        if (this._loadingCapped) {
            return false;
        }
        this._loadingCapped = true;
        Utils.noAwait(async () => {
            await Utils.delay(30);
            this._loadingCapped = false;
        });
        return await this._loader.process(
            () => this._bridge.call(this._messageHelper.load),
            result => {
                // OK.
            });
    }

    public show(): Promise<FullScreenAdResult> {
        return this._displayer.process(
            () => this._bridge.call(this._messageHelper.show),
            result => {
                // OK.
            });
    }

    private onLoaded(): void {
        if (this._loader.isProcessing) {
            this._loader.resolve(true);
        } else {
            // Assert.IsTrue(false);
        }
        this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded());
    }

    private onFailedToLoad(message: string): void {
        if (this._loader.isProcessing) {
            this._loader.resolve(false);
        } else {
            // Assert.
        }
    }

    private onFailedToShow(message: string): void {
        if (this._displayer.isProcessing) {
            this._displayer.resolve(FullScreenAdResult.Failed);
        } else {
            // Assert.
        }
    }

    private onClicked(): void {
        this.dispatchEvent(observer => observer.onClicked && observer.onClicked());
    }

    private onClosed(rewarded: boolean): void {
        if (this._displayer.isProcessing) {
            this._displayer.resolve(rewarded
                ? FullScreenAdResult.Completed
                : FullScreenAdResult.Canceled);
        } else {
            // Assert.
        }
    }
}